panic: invalid compaction skipped

- JuiceFS version 1.2.0+2024-06-18.873c47b
- BadgerDB

### Syslog
```
juicefs[707577]: 2024/08/31 08:06:48.386811 badger[707577] <INFO>: Set nextTxnTs to 7718283 [logger.go:46]
juicefs[707577]: 2024/08/31 08:06:48.386881 badger[707577] <INFO>: Deleting empty file: /mnt/volume/badger/003886.vlog [logger.go:46]
juicefs[707577]: 2024/08/31 08:06:48.416101 juicefs[707577] <INFO>: Data use s3://ceezy-lpbdocs/juicefs/(encrypted) [mount.go:629]
juicefs[707577]: 2024/08/31 08:06:48.416156 juicefs[707577] <INFO>: JuiceFS version 1.2.0+2024-06-18.873c47b [mount.go:659]
juicefs[707577]: 2024/08/31 08:06:48.419062 juicefs[707577] <INFO>: Disk cache (/var/cache/juicefs/7af5d98d-f5d6-4f36-8e38-648ca7854b26/): capacity (100 GiB), free ratio 20%, used ratio - [space 67%, inode 4.3%], max pending pages 15 [disk_cache.go:144]
juicefs[707577]: 2024/08/31 08:06:48.419228 juicefs[707577] <INFO>: delay uploading by 1m30s [cached_store.go:833]
juicefs[707577]: 2024/08/31 08:06:48.419949 juicefs[707577] <INFO>: Create session 3804 OK with version: 1.2.0+2024-06-18.873c47b [base.go:398]
juicefs[707577]: 2024/08/31 08:06:48.421802 juicefs[707577] <INFO>: Prometheus metrics listening on 127.0.0.1:9567 [mount.go:135]
juicefs[707577]: 2024/08/31 08:06:48.422439 juicefs[707577] <INFO>: Mounting volume juicefs at /var/data ... [mount_unix.go:807]
juicefs[707577]: 2024/08/31 08:06:48.422448 juicefs[707577] I | Unimplemented opcode POLL
juicefs[412649]: 2024/08/31 08:06:48.424462 juicefs[412649] <INFO>: watching /var/data, pid 707577 [mount_unix.go:152]
juicefs[707577]: 2024-08-31 08:06:51.426084 I | FUSE: interrupt request 59129999
juicefs[707577]: 2024-08-31 08:06:51.426143 I | FUSE: interrupt request 59129998
juicefs[707577]: 2024-08-31 08:06:51.426151 I | FUSE: interrupt request 59129997
juicefs[707577]: 2024-08-31 08:06:51.426166 I | FUSE: interrupt request 59129996
juicefs[707577]: panic: invalid compaction skipped 1, pos 0, size 47100348; slices:
juicefs[707577]: {id:2086138 size:47100348 off:0 len:47100348 pos:0 left:<nil> right:<nil>}
juicefs[707577]: {id:0 size:0 off:0 len:47100348 pos:0 left:<nil> right:<nil>}
juicefs[707577]: {id:2086138 size:47100348 off:0 len:47100348 pos:0 left:<nil> right:<nil>}
juicefs[707577]: {id:0 size:0 off:0 len:47100348 pos:0 left:<nil> right:<nil>}
juicefs[707577]: {id:2086138 size:47100348 off:0 len:47100348 pos:0 left:<nil> right:<nil>}
juicefs[707577]: {id:0 size:0 off:0 len:47100348 pos:0 left:<nil> right:<nil>}
juicefs[707577]: {id:2086138 size:47100348 off:0 len:47100348 pos:0 left:<nil> right:<nil>}
juicefs[707577]: {id:0 size:0 off:0 len:47100348 pos:0 left:<nil> right:<nil>}
juicefs[707577]: {id:2086138 size:47100348 off:0 len:47100348 pos:0 left:<nil> right:<nil>}
juicefs[707577]: goroutine 373 [running]:
juicefs[707577]: github.com/juicedata/juicefs/pkg/meta.(*baseMeta).compactChunk(0xc00076a500, 0x196e3a, 0x0, 0x0, 0x0)
juicefs[707577]: #011/go/src/github.com/juicedata/juicefs/pkg/meta/base.go:2056 +0x10c5
juicefs[707577]: created by github.com/juicedata/juicefs/pkg/meta.(*baseMeta).Read in goroutine 370
juicefs[707577]: #011/go/src/github.com/juicedata/juicefs/pkg/meta/base.go:1428 +0x4b1
```

### FSCK
```
# juicefs fsck badger:///mnt/volume/badger

2024/08/31 08:41:57.426628 juicefs[2126] <INFO>: Meta address: badger:///mnt/volume/badger [interface.go:504]
2024/08/31 08:41:57.441174 badger[2126] <INFO>: All 30 tables opened in 1ms [logger.go:46]
2024/08/31 08:41:57.443922 badger[2126] <INFO>: Discard stats nextEmptySlot: 0 [logger.go:46]
2024/08/31 08:41:57.443999 badger[2126] <INFO>: Set nextTxnTs to 7722750 [logger.go:46]
2024/08/31 08:41:57.444078 badger[2126] <INFO>: Deleting empty file: /mnt/volume/badger/004020.vlog [logger.go:46]
2024/08/31 08:41:57.445901 juicefs[2126] <INFO>: Data use s3://ceezy-lpbdocs/juicefs/(encrypted) [fsck.go:108]
  Found blocks: 829820                            3907.0/s
  Found blocks: 919.6 GiB (987360329395 Bytes)    4.3 GiB/s
 Listed slices: 639280                            217958.9/s
 Deleted files: 0                                 0.0/s
Skipped slices: 0                                 0.0/s
Scanned slices: 639280/639280 [==============================================================]  541570.8/s used: 1.180426713s
Scanned slices: 1.1 TiB   (1156702505863 Bytes)   913.3 GiB/s
   Lost blocks: 0                                 0.0/s
   Lost blocks: 0.0 b     (0 Bytes)               0.0 b/s
```

### GC Delete
```
# juicefs gc --delete badger:///mnt/volume/badger

2024/08/31 08:49:39.367312 juicefs[2492] <INFO>: Meta address: badger:///mnt/volume/badger [interface.go:504]
2024/08/31 08:49:39.382184 badger[2492] <INFO>: All 27 tables opened in 2ms [logger.go:46]
2024/08/31 08:49:39.384438 badger[2492] <INFO>: Discard stats nextEmptySlot: 0 [logger.go:46]
2024/08/31 08:49:39.384499 badger[2492] <INFO>: Set nextTxnTs to 7723981 [logger.go:46]
2024/08/31 08:49:39.384603 badger[2492] <INFO>: Deleting empty file: /mnt/volume/badger/004023.vlog [logger.go:46]
2024/08/31 08:49:39.386146 juicefs[2492] <INFO>: Data use s3://ceezy-lpbdocs/juicefs/(encrypted) [gc.go:101]
Cleaned pending slices: 5                                0.0/s
 Pending deleted files: 0                                0.0/s
  Pending deleted data: 0.0 b     (0 Bytes)              0.0 b/s
 Cleaned pending files: 0                                0.0/s
  Cleaned pending data: 0.0 b     (0 Bytes)              0.0 b/s
         Cleaned trash: 0                                0.0/s
Cleaned detached nodes: 0                                0.0/s
         Listed slices: 638912                           3017.1/s
          Trash slices: 0                                0.0/s
            Trash data: 0.0 b     (0 Bytes)              0.0 b/s
  Cleaned trash slices: 0                                0.0/s
    Cleaned trash data: 0.0 b     (0 Bytes)              0.0 b/s
       Scanned objects: 829530/829530 [=====================================================]  3974.6/s used: 3m28.708220235s
         Valid objects: 829142                           3972.7/s
            Valid data: 919.3 GiB (987080678577 Bytes)   4.4 GiB/s
     Compacted objects: 0                                0.0/s
        Compacted data: 0.0 b     (0 Bytes)              0.0 b/s
        Leaked objects: 8                                0.0/s
           Leaked data: 19.5 MiB  (20397109 Bytes)       95.4 KiB/s
       Skipped objects: 380                              1.8/s
          Skipped data: 284.3 MiB (298067987 Bytes)      1.4 MiB/s
2024/08/31 08:53:11.184788 juicefs[2492] <INFO>: scanned 829530 objects, 829142 valid, 0 compacted (0 bytes), 8 leaked (20397109 bytes), 0 delslices (0 bytes), 0 delfiles (0 bytes), 380 skipped (298067987 bytes) [gc.go:379]
```

### GC Compact
```
# juicefs gc --compact badger:///mnt/volume/badger

2024/08/31 08:48:42.136179 juicefs[2458] <INFO>: Meta address: badger:///mnt/volume/badger [interface.go:504]
2024/08/31 08:48:42.155323 badger[2458] <INFO>: All 26 tables opened in 1ms [logger.go:46]
2024/08/31 08:48:42.157634 badger[2458] <INFO>: Discard stats nextEmptySlot: 0 [logger.go:46]
2024/08/31 08:48:42.157699 badger[2458] <INFO>: Set nextTxnTs to 7723980 [logger.go:46]
2024/08/31 08:48:42.157819 badger[2458] <INFO>: Deleting empty file: /mnt/volume/badger/004022.vlog [logger.go:46]
2024/08/31 08:48:42.159676 juicefs[2458] <INFO>: Data use s3://ceezy-lpbdocs/juicefs/(encrypted) [gc.go:101]
Cleaned pending slices: 0                 0.0/s                                   ⠦
 Pending deleted files: 0                 0.0/s
  Pending deleted data: 0.0 b (0 Bytes)   0.0 b/s
 Cleaned pending files: 0                 0.0/s
  Cleaned pending data: 0.0 b (0 Bytes)   0.0 b/s
      Compacted chunks: 677/679 [==============================================================]  1477.0/s ETA: 0s
      Compacted slices: 0                 0.0/s                                   ⠸
        Compacted data: 0.0 b (0 Bytes)   0.0 b/s                                 ⠸
panic: invalid compaction skipped 1, pos 0, size 47100348; slices:
{id:2086138 size:47100348 off:0 len:47100348 pos:0 left:<nil> right:<nil>}
{id:0 size:0 off:0 len:47100348 pos:0 left:<nil> right:<nil>}
{id:2086138 size:47100348 off:0 len:47100348 pos:0 left:<nil> right:<nil>}
{id:0 size:0 off:0 len:47100348 pos:0 left:<nil> right:<nil>}
{id:2086138 size:47100348 off:0 len:47100348 pos:0 left:<nil> right:<nil>}
{id:0 size:0 off:0 len:47100348 pos:0 left:<nil> right:<nil>}
{id:2086138 size:47100348 off:0 len:47100348 pos:0 left:<nil> right:<nil>}
{id:0 size:0 off:0 len:47100348 pos:0 left:<nil> right:<nil>}
{id:2086138 size:47100348 off:0 len:47100348 pos:0 left:<nil> right:<nil>}

goroutine 142 [running]:
github.com/juicedata/juicefs/pkg/meta.(*baseMeta).compactChunk(0xc000782500, 0x196e3a, 0x0, 0x0, 0x1)
        /go/src/github.com/juicedata/juicefs/pkg/meta/base.go:2056 +0x10c5
github.com/juicedata/juicefs/pkg/meta.(*baseMeta).CompactAll.func1()
        /go/src/github.com/juicedata/juicefs/pkg/meta/base.go:1991 +0x1d5
created by github.com/juicedata/juicefs/pkg/meta.(*baseMeta).CompactAll in goroutine 1
        /go/src/github.com/juicedata/juicefs/pkg/meta/base.go:1988 +0x7a
```

--- a/pkg/meta/base_test.go
+++ b/pkg/meta/base_test.go
@@ -1573,6 +1573,21 @@ func testCompaction(t *testing.T, m Meta, trash bool) {
 		t.Fatalf("inode %d should be compacted, but have %d slices, size %d,%d,%d",
 			inode, len(slices), slices[0].Len, slices[1].Len, slices[2].Len)
 	}
+
+	m.NewSlice(ctx, &sliceId)
+	_ = m.Write(ctx, inode, 3, 0, Slice{Id: sliceId, Size: 2338508, Len: 2338508}, time.Now())
+	_ = m.CopyFileRange(ctx, inode, 3*ChunkSize, inode, 4*ChunkSize, 2338508, 0, nil, nil)
+	_ = m.Fallocate(ctx, inode, fallocZeroRange, 4*ChunkSize, ChunkSize, nil)
+	_ = m.CopyFileRange(ctx, inode, 3*ChunkSize, inode, 4*ChunkSize, 2338508, 0, nil, nil)
+	if c, ok := m.(compactor); ok {
+		c.compactChunk(inode, 4, false, true)
+	}
+	if st := m.Read(ctx, inode, 4, &slices); st != 0 {
+		t.Fatalf("read inode %d chunk 4: %s", inode, st)
+	}
+	if len(slices) != 1 || slices[0].Len != 2338508 {
+		t.Fatalf("inode %d should be compacted, but have %d slices, size %d", inode, len(slices), slices[0].Len)
+	}
 }
 
 func testConcurrentWrite(t *testing.T, m Meta) {

--- a/pkg/meta/slice.go
+++ b/pkg/meta/slice.go
@@ -183,6 +183,7 @@ func compactChunk(ss []*slice) (uint32, uint32, []Slice) {
 func skipSome(chunk []*slice) int {
 	var skipped int
 	var total = len(chunk)
+OUT:
 	for skipped < total {
 		ss := chunk[skipped:]
 		pos, size, c := compactChunk(ss)
@@ -197,6 +198,11 @@ func skipSome(chunk []*slice) int {
 		if !isFirst(pos, c[0]) {
 			// it's not the first slice, compact it
 			break
+		}
+		for _, s := range ss[1:] {
+			if *s == *first {
+				break OUT
+			}
 		}
 		skipped++
 	}

